isolateEmulatorThread: Add full-pcpu-only support

RamLavi opened the pull request:
**What this PR does / why we need it**:
This PR add an annotation related to the `isolateEmulatorThread` feature, in order to avoid getting SMT alignment error by kubelet on the virt-launcher pod:
```
SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
```

The conditions needed in order to get this error:
- The node has hyperthreading (SMT) enabled.
- Kubelet’s CPUManager is set to static - full-pcpus-only.
- Kubevirt CR's defaultRuntimeClass is set to a high performance runtimeClass.
- VM is set with dedicatedCPUPlacement and IsolateEmulatorThread enabled.
- VM requests an even number of CPUs.

The meaning of this is that instead of IsolateEmulatorThread adding an extra CPU to the virt-launcher pod, it will add two CPUs when needed in order to get a total number of even CPUs.

Since this error is a direct result of adding the emulatorCPU by the IsolateEmulatorThread feature - the fix is currently needed on kubevirt context.

Note: this PR implements the first phase of the [design proposal](https://github.com/kubevirt/community/pull/247). 
Future PR(s) will utilize the extra CPU to the housekeeping cgroup.


**Which issue(s) this PR fixes**:
Fixes [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103)

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [x] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Fixes SMT Alignment Error in virt-launcher pod by optimizing isolateEmulatorThread feature (BZ#2228103).
```

RamLavi commented:
/test pull-kubevirt-generate 
/test pull-kubevirt-unit-test
/test pull-kubevirt-verify-go-mod

RamLavi commented:
/test pull-kubevirt-generate
/test pull-kubevirt-unit-test
/test pull-kubevirt-verify-go-mod

orelmisan reviewed (commented):
Thank you for the PR @RamLavi.

PR description:
Please consider adding the list of conditions that leads to the problem being solved (as are present in the design PR):
* The node has hyperthreading (SMT) enabled.
* Kubelet’s CPUManager is set to static - full-pcpus-only.
* Kubevirt CR's defaultRuntimeClass is set to a high performance runtimeClass.
* VM is set with `dedicatedCPUPlacement` and `IsolateEmulatorThread` enabled.
* VM requests an even number of CPUs.

Please consider stating that this is the first of part of the solution, which enables spinning up VMs that meet the above conditions, while not using the second extra CPU.

RamLavi commented:
> Thank you for the PR @RamLavi.
> 
> PR description: Please consider adding the list of conditions that leads to the problem being solved (as are present in the design PR):
> 
> * The node has hyperthreading (SMT) enabled.
> * Kubelet’s CPUManager is set to static - full-pcpus-only.
> * Kubevirt CR's defaultRuntimeClass is set to a high performance runtimeClass.
> * VM is set with `dedicatedCPUPlacement` and `IsolateEmulatorThread` enabled.
> * VM requests an even number of CPUs.
> 
> Please consider stating that this is the first of part of the solution, which enables spinning up VMs that meet the above conditions, while not using the second extra CPU.

Thanks for the comments! I changed accordingly, could you re-review?

RamLavi commented:
/test pull-kubevirt-generate
/test pull-kubevirt-unit-test
/test pull-kubevirt-verify-go-mod

RamLavi commented:
[Change](https://github.com/kubevirt/kubevirt/compare/f3604cfdec773cc8454887d4a14c44f45471ed2c..3d6cf6c114b70e6261fff0b5fbefce7c5e0d9b1f): rebase.

/test pull-kubevirt-generate
/test pull-kubevirt-unit-test
/test pull-kubevirt-verify-go-mod

orelmisan commented on pkg/virt-controller/services/renderresources.go:
```diff
@@ -199,7 +199,7 @@ func WithAutoMemoryLimits(namespace string, namespaceStore cache.Store) Resource
 }
 
-func WithCPUPinning(cpu *v1.CPU) ResourceRendererOption {
+func WithCPUPinning(cpu *v1.CPU, cpuManagerPolicyBetaOptionsAnnotationValue string) ResourceRendererOption {
```
Please consider changing the type here to the enum, and adding an `Unknown` or `Unset` value:
```suggestion
func WithCPUPinning(cpu *v1.CPU, cpuManagerPolicyBetaOptions v1.CPUManagerPolicyBetaOptions) ResourceRendererOption {
```

orelmisan replied:
The advantage is that it limits the number of possible inputs to the string constants comprising the `CPUManagerPolicyBetaOptions` type (and not any string value).

Indeed, some function higher in the call stack will have to perform this conversion.

RamLavi replied:
DONE

orelmisan commented on pkg/virt-controller/services/renderresources.go:
```diff
@@ -212,14 +212,20 @@ func WithCPUPinning(cpu *v1.CPU) ResourceRendererOption {
 		}
 
-		// allocate 1 more pcpu if IsolateEmulatorThread request
+		// allocate pcpus for emulatorThread if IsolateEmulatorThread is requested
```
```suggestion
		// allocate one or two pcpus for emulatorThread if IsolateEmulatorThread is requested
```

RamLavi replied:
I prefer to be generic. comment tend to get stale, keeping generally "plural" without being explicit will help keeping it true.

orelmisan commented on pkg/virt-controller/services/renderresources.go:
```diff
@@ -212,14 +212,20 @@ func WithCPUPinning(cpu *v1.CPU) ResourceRendererOption {
 		if cpu.IsolateEmulatorThread {
-			emulatorThreadCPU := resource.NewQuantity(1, resource.BinarySI)
+			const emulatorThreadCPU = 1
+			emulatorThreadCPUs := resource.NewQuantity(emulatorThreadCPU, resource.BinarySI)
```
AFAIK `resource.BinarySI` is used to define resources of type memory or storage.
I think it should be `resource.DecimalSI`.

RamLavi replied:
It's not directly related to this change, Can we pin this for a separate PR?

RamLavi replied:
> It's not directly related to this change. Can we pin this for a separate PR?

orelmisan replied:
Because the values are one and two, it doesn't seem to matter, but in the tests there are decimal numbers being used `1000m` etc so it is a bit confusing IMO.

RamLavi replied:
I agree, but it's a discussion that deserves a separate PR IMO.

orelmisan commented on pkg/virt-controller/services/renderresources.go:
```diff
@@ -287,6 +293,14 @@ func copyResources(srcResources, dstResources k8sv1.ResourceList) {
 	}
 }
 
+func getEmulatorThreadCPUsSMTAlignment(cpuLimits, emulatorThreadCPUsBeforeAlignment resource.Quantity) *resource.Quantity {
+	const threadsPerCore = 2
+	cpuLimitsBeforeAlignment := cpuLimits.DeepCopy()
+	cpuLimitsBeforeAlignment.Add(emulatorThreadCPUsBeforeAlignment)
+	fullPCPUAlignment := cpuLimitsBeforeAlignment.Value() % threadsPerCore
+	return resource.NewQuantity(fullPCPUAlignment, resource.BinarySI)
+}
```
Please consider using this logic inline.

RamLavi replied:
Can you elaborate why? I don't think this implementation detail should be seen outside of this function.

orelmisan replied:
Because the logic is quite simple:
If isolatedEmulatorThread is required - KubeVirt requests one extra CPU.
If CPUManager option is set to fullPCPUs and the total number of requested CPUs is odd - add another CPU.

IMO adding another function makes this flow less readable.

RamLavi replied:
The logic I inserted here is a bit "math"ish. it has the benefit of being true with every natural value of threadsPerCore we may use. The downside is that "math" equations are less readable for code reviewers. 
I balance this by hiding the simple math equation as an implementation detail in functions.

orelmisan replied:
Please consider the following implementation (not tested):
```go
		if cpu.IsolateEmulatorThread {
			extraCPUs := resource.NewQuantity(1, resource.BinarySI)

			limits := renderer.calculatedLimits[k8sv1.ResourceCPU]
			if cpuManagerPolicyBetaOption == v1.CPUManagerPolicyBetaOptionFullpCPUsOnly && limits.Value() % 2 == 0{
				extraCPUs = resource.NewQuantity(2, resource.BinarySI)
			} 
			
			limits.Add(*extraCPUs)
			renderer.vmLimits[k8sv1.ResourceCPU] = limits
			
			if cpuRequest, ok := renderer.vmRequests[k8sv1.ResourceCPU]; ok {
				cpuRequest.Add(*extraCPUs)
				renderer.vmRequests[k8sv1.ResourceCPU] = cpuRequest
			}
		}
```

RamLavi replied:
while the current assumption is that a core can hold only 2 threads, technology evolves, and this assumption may change in the future. By using a math equation that always holds true (in our case case, `fullPCPUAlignment := cpuLimitsBeforeAlignment.Value() % threadsPerCore`), we future-proof the code (or at least make it easier to our future-selves) and avoid the need to add additional conditionals if the core-thread relationship changes. 
This approach enhances code maintainability and readability in the long run.

orelmisan replied:
> while the current assumption is that a core can hold only 2 threads, technology evolves, and this assumption may change in the future. By using a math equation that always holds true (in our case case, `fullPCPUAlignment := cpuLimitsBeforeAlignment.Value() % threadsPerCore`), we future-proof the code (or at least make it easier to our future-selves) and avoid the need to add additional conditionals if the core-thread relationship changes. This approach enhances code maintainability and readability in the long run.

The suggested implementation above, adds two extra CPUs only in case the VM's CPU count is even (and the correct CPUManager option is selected).

The assumption is that in order to be SMT aligned, the total number of requested CPUs should be even.

This also works in case the user requested an odd number of CPUs - KubeVirt will add a single CPU which will cause the total CPU count on the virt-launcher pod to be even.

RamLavi replied:
talked offline - agreed to use your suggestion as it better fits the design.

orelmisan replied:
Thank you for considering and accepting this suggestion.
Please note that in the original suggestion, the second part of the condition is meant to check if the total CPU count is even.

I suggest to remove the `fullPCPUsPerCore` constant, as it changes the intention to something different than the parity.

orelmisan replied:
The second part of the condition could potentially be replaced with a function `isEven()` in case you find the modulo is not clear enough.

RamLavi replied:
DONE

orelmisan commented on pkg/virt-controller/services/renderresources.go:
```diff
@@ -212,14 +212,20 @@ func WithCPUPinning(cpu *v1.CPU) ResourceRendererOption {
+			emulatorThreadCPUs := resource.NewQuantity(emulatorThreadCPU, resource.BinarySI)
 			limits := renderer.calculatedLimits[k8sv1.ResourceCPU]
-			limits.Add(*emulatorThreadCPU)
+			if v1.CPUManagerPolicyBetaOptions(cpuManagerPolicyBetaOptionsAnnotationValue) == v1.FullpCPUsOnly {
```
In case the function's input param is changed to `v1.CPUManagerPolicyBetaOptions`:
```suggestion
if cpuManagerPolicyBetaOption == v1.FullpCPUsOnly {
```

RamLavi replied:
pending your answer on the related comment.

RamLavi replied:
DONE

RamLavi commented on staging/src/kubevirt.io/api/core/v1/types.go:
```diff
@@ -2211,6 +2214,13 @@ type VirtualMachineMemoryDumpRequest struct {
+
+const (
+	// full-pcpus-only static policy allows only full physical CPUs to be scheduled to a pod
+	FullpCPUsOnly CPUManagerPolicyBetaOptions = "full-pcpus-only"
```
DONE

RamLavi commented on pkg/virt-config/feature-gates.go:
```diff
@@ -81,6 +81,8 @@ const (
 	//
 	// CommonInstancetypesDeploymentGate enables the deployment of common-instancetypes by virt-operator
 	CommonInstancetypesDeploymentGate = "CommonInstancetypesDeploymentGate"
+	// CPUManagerPolicyBetaOptionsGate allows conforming VM to static CPU-Manager Policies.
```
DONE.

RamLavi commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:
```diff
@@ -1757,6 +1757,15 @@ func ValidateVirtualMachineInstanceMetadata(field *k8sfield.Path, metadata *meta
 		}
 	}
 
+	if _, exists := annotations[v1.CPUManagerPolicyBetaOptionsAnnotation]; exists && !config.CPUManagerPolicyBetaOptionsEnabled() {
```
DONE

RamLavi commented:
@orelmisan addressed comments [here](https://github.com/kubevirt/kubevirt/compare/3d6cf6c114b70e6261fff0b5fbefce7c5e0d9b1f..d87c1596503e422ef3bca3632e4269b1609746ba), and also commented back on some. could you please re-review?

RamLavi commented:
Change: [Rebase](https://github.com/kubevirt/kubevirt/compare/d87c1596503e422ef3bca3632e4269b1609746ba..17675a957671b9ad2461df58a79d6c80651a174f)

RamLavi commented:
[Change](https://github.com/kubevirt/kubevirt/compare/ba8bc92855add9707dd92fa6c16273cf20d7f047..b21bc374d507f30f9731a3dba2db1831d6a8d394): Simplified extra CPU logic to increase readability and conform to the design doc.

orelmisan commented on pkg/virt-controller/services/renderresources_test.go:
```diff
@@ -195,42 +196,80 @@ var _ = Describe("Resource pod spec renderer", func() {
-		When("an isolated emulator thread is requested", func() {
-			cpuIsolatedEmulatorThreadOverhead := resource.MustParse("1000m")
-			userSpecifiedCPURequest := kubev1.ResourceList{kubev1.ResourceCPU: userCPURequest}
+		When("and isolated emulator thread is requested", func() {
```
nit:
```suggestion
		When("an isolated emulator thread is requested", func() {
```

RamLavi replied:
DONE

orelmisan commented on pkg/virt-controller/services/template_test.go:
```diff
@@ -1716,9 +1720,13 @@ var _ = Describe("Template", func() {
 				Expect(pod.Spec.Containers[0].Resources.Limits.Cpu().Cmp(cpu)).To(BeZero())
-			})
+			},
+				Entry("no annotation added, should allocate one extra emulator CPU", map[string]string{}, uint32(2), "3"),
```
Please consider adding an additional entry without the annotation and with an odd number of CPUs as input.

RamLavi replied:
sure. DONE.

RamLavi commented on pkg/virt-controller/services/renderresources_test.go:
```diff
@@ -195,42 +196,80 @@ var _ = Describe("Resource pod spec renderer", func() {
-			cpuIsolatedEmulatorThreadOverhead := resource.MustParse("1000m")
-			userSpecifiedCPURequest := kubev1.ResourceList{kubev1.ResourceCPU: userCPURequest}
+		When("and isolated emulator thread is requested", func() {
+			When("and full-pcpu-only mode is disabled", func() {
```
DONE

RamLavi commented on pkg/virt-controller/services/renderresources_test.go:
```diff
@@ -195,42 +196,80 @@ var _ = Describe("Resource pod spec renderer", func() {
-				Entry("request and limits set by the user", true),
-			)
+			})
+			When("and full-pcpu-only mode is enabled", func() {
```
DONE

RamLavi commented on pkg/virt-controller/services/template.go:
```diff
@@ -1432,7 +1432,8 @@ func (t *templateService) VMIResourcePredicates(vmi *v1.VirtualMachineInstance,
 		resourceRules: []VMIResourceRule{
-			NewVMIResourceRule(doesVMIRequireDedicatedCPU, WithCPUPinning(vmi.Spec.Domain.CPU)),
+			NewVMIResourceRule(doesVMIRequireDedicatedCPU,
+				WithCPUPinning(vmi.Spec.Domain.CPU, v1.CPUManagerPolicyBetaOptions(vmi.Annotations[v1.CPUManagerPolicyBetaOptionsAnnotation]))),
```
sure. DONE

RamLavi commented:
[Change](https://github.com/kubevirt/kubevirt/compare/1cbb802d24f48da233e69e933fbbbb6223842179..a571f8fad1fff0ee401d1df924e132429d8601ad): fixes @orelmisan

RamLavi commented:
[Change](https://github.com/kubevirt/kubevirt/compare/a571f8fad1fff0ee401d1df924e132429d8601ad..3fc9c3fcbe6fa46eccc1d03aad6ae511e5e6c844): Rebase fix

RamLavi commented on pkg/virt-controller/services/renderresources_test.go:
```diff
@@ -195,42 +196,80 @@ var _ = Describe("Resource pod spec renderer", func() {
-						Cores:                 5,
-						IsolateEmulatorThread: true,
-					}),
+			When("full-pcpu-only mode is disabled", func() {
```
yes, but I'm not sure it made it prettier (see [here](https://github.com/kubevirt/kubevirt/compare/3fc9c3fcbe6fa46eccc1d03aad6ae511e5e6c844..2e5cb47f337d47dfd6504473ec1365e159665d20)). tell me what do you think.

RamLavi commented:
[Change](https://github.com/kubevirt/kubevirt/compare/3fc9c3fcbe6fa46eccc1d03aad6ae511e5e6c844..2e5cb47f337d47dfd6504473ec1365e159665d20): Combined the unit tests to the same context on renderresources_tests

RamLavi commented:
/retest-required

RamLavi commented:
[Change](https://github.com/kubevirt/kubevirt/compare/3fc9c3fcbe6fa46eccc1d03aad6ae511e5e6c844..fb7d43fd085dccf62c0c51050eddb8a259142b99): Combined the unit tests to the same context on renderresources_tests

RamLavi commented:
[Change](https://github.com/kubevirt/kubevirt/compare/fb7d43fd085dccf62c0c51050eddb8a259142b99..f8aa00822498f801d095b1f31ad264b78d5608f2): commit typos

orelmisan reviewed (approved):
Thank you

RamLavi commented:
/retest-required

jean-edouard commented:
/approve
Please add a paragraph about this in the user-guide, around https://kubevirt.io/user-guide/virtual_machines/dedicated_cpu_resources/#requesting-dedicated-cpu-for-qemu-emulator
Thank you!

RamLavi commented:
/cherry-pick v1.1.0

vladikr commented:
Hey @RamLavi @jean-edouard @orelmisan @xpivarc ,

Sorry that I'm late to review this, however, I'm missing here any configuration of the emulator thread on the VM side, nor this affects the housekeeping cgrpoup.

What I see in this PR is only an increase of CPU resources on the virt-launcher's compute container, but I would expect that the emulator thread would also be pinned to the additional CPU in 
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go#L362
This means that the logic in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go#L458-L465
would need to change the way that we will allocate 2 threads when needed. `emulatorThread` would need to be a slice.

Once the converter part is adjusted, the virt-handler will be able to correctly create the housekeeping cgroup in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm.go#L2828

We need a follow up PR.

RamLavi commented:
> Hey @RamLavi @jean-edouard @orelmisan @xpivarc ,
> 
> Sorry that I'm late to review this, however, I'm missing here any configuration of the emulator thread on the VM side, nor this affects the housekeeping cgrpoup.
> 
> What I see in this PR is only an increase of CPU resources on the virt-launcher's compute container, but I would expect that the emulator thread would also be pinned to the additional CPU in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go#L362 This means that the logic in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go#L458-L465 would need to change the way that we will allocate 2 threads when needed. `emulatorThread` would need to be a slice.
> 
> Once the converter part is adjusted, the virt-handler will be able to correctly create the housekeeping cgroup in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm.go#L2828
> 
> We need a follow up PR.

Hey  @vladikr 
You are correct. 
We decided to break the design into 2 [phases](https://github.com/kubevirt/community/pull/247/files#diff-f4c36916d38420c6a9bc23cfbffef0d45f1192e3ca947b587fbcab669d40d4ffR155). The first phase is implemented on this PR, and the other will be set on another, to reduce the complexity of each PR.

RamLavi commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:
```diff
@@ -1717,6 +1717,15 @@ func ValidateVirtualMachineInstanceMetadata(field *k8sfield.Path, metadata *meta
+	if _, exists := annotations[v1.CPUManagerPolicyBetaOptionsAnnotation]; exists && !config.CPUManagerPolicyBetaOptionsEnabled() {
+		causes = append(causes, metav1.StatusCause{
+			Type: metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("CPUManagerPolicyBetaOptions feature gate is not enabled in kubevirt-config, invalid entry %s",
```
not sure I got what you mean. Can you be more specific?

xpivarc replied:
The path below should be enough, rather than embedding it into the message.

RamLavi replied:
I see what you mean. Will fix in a followup PR

vladikr commented:
> > Hey @RamLavi @jean-edouard @orelmisan @xpivarc ,
> > Sorry that I'm late to review this, however, I'm missing here any configuration of the emulator thread on the VM side, nor this affects the housekeeping cgrpoup.
> > What I see in this PR is only an increase of CPU resources on the virt-launcher's compute container, but I would expect that the emulator thread would also be pinned to the additional CPU in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go#L362 This means that the logic in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go#L458-L465 would need to change the way that we will allocate 2 threads when needed. `emulatorThread` would need to be a slice.
> > Once the converter part is adjusted, the virt-handler will be able to correctly create the housekeeping cgroup in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm.go#L2828
> > We need a follow up PR.
> 
> Hey @vladikr You are correct. We decided to break the design into 2 [phases](https://github.com/kubevirt/community/pull/247/files#diff-f4c36916d38420c6a9bc23cfbffef0d45f1192e3ca947b587fbcab669d40d4ffR155). The first phase is implemented on this PR, and the other will be set on another, to reduce the complexity of each PR.

Hmm. I think I missed that. However, without the second part the feature is incomplete. 
I expect both PRs to be present in each version.
Is there a PR for the second part already?

RamLavi commented:
no, but as soon as it's done I'll link it here.

RamLavi commented:
> > > Hey @RamLavi @jean-edouard @orelmisan @xpivarc ,
> > > Sorry that I'm late to review this, however, I'm missing here any configuration of the emulator thread on the VM side, nor this affects the housekeeping cgrpoup.
> > > What I see in this PR is only an increase of CPU resources on the virt-launcher's compute container, but I would expect that the emulator thread would also be pinned to the additional CPU in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go#L362 This means that the logic in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go#L458-L465 would need to change the way that we will allocate 2 threads when needed. `emulatorThread` would need to be a slice.
> > > Once the converter part is adjusted, the virt-handler will be able to correctly create the housekeeping cgroup in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm.go#L2828
> > > We need a follow up PR.
> > 
> > 
> > Hey @vladikr You are correct. We decided to break the design into 2 [phases](https://github.com/kubevirt/community/pull/247/files#diff-f4c36916d38420c6a9bc23cfbffef0d45f1192e3ca947b587fbcab669d40d4ffR155). The first phase is implemented on this PR, and the other will be set on another, to reduce the complexity of each PR.
> 
> Hmm. I think I missed that. However, without the second part the feature is incomplete. I expect both PRs to be present in each version. Is there a PR for the second part already?

I agree 100%. I still don't have a PR but I'll link it as soon it I push it

RamLavi commented:
> > > > Hey @RamLavi @jean-edouard @orelmisan @xpivarc ,
> > > > Sorry that I'm late to review this, however, I'm missing here any configuration of the emulator thread on the VM side, nor this affects the housekeeping cgrpoup.
> > > > What I see in this PR is only an increase of CPU resources on the virt-launcher's compute container, but I would expect that the emulator thread would also be pinned to the additional CPU in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go#L362 This means that the logic in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go#L458-L465 would need to change the way that we will allocate 2 threads when needed. `emulatorThread` would need to be a slice.
> > > > Once the converter part is adjusted, the virt-handler will be able to correctly create the housekeeping cgroup in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm.go#L2828
> > > > We need a follow up PR.
> > > 
> > > 
> > > Hey @vladikr You are correct. We decided to break the design into 2 [phases](https://github.com/kubevirt/community/pull/247/files#diff-f4c36916d38420c6a9bc23cfbffef0d45f1192e3ca947b587fbcab669d40d4ffR155). The first phase is implemented on this PR, and the other will be set on another, to reduce the complexity of each PR.
> > 
> > 
> > Hmm. I think I missed that. However, without the second part the feature is incomplete. I expect both PRs to be present in each version. Is there a PR for the second part already?
> 
> I agree 100%. I still don't have a PR but I'll link it as soon it I push it

@vladikr FYI https://github.com/kubevirt/kubevirt/pull/10783
